File: mojo/bindings/js/unicode.cc
// Copyright 2014 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "mojo/bindings/js/unicode.h"

#include "gin/arguments.h"
#include "gin/array_buffer.h"
#include "gin/object_template_builder.h"
#include "gin/per_isolate_data.h"
#include "gin/public/wrapper_info.h"

namespace mojo {
namespace js {

namespace {

v8::Handle<v8::Value> DecodeUtf8String(const gin::Arguments& args,
                                       const gin::ArrayBufferView& buffer) {
  assert(static_cast<int>(buffer.num_bytes()) >= 0);
  return v8::String::NewFromUtf8(args.isolate(),
                                 reinterpret_cast<char*>(buffer.bytes()),
                                 v8::String::kNormalString,
                                 static_cast<int>(buffer.num_bytes()));
}

int EncodeUtf8String(const gin::Arguments& args,
                     v8::Handle<v8::Value> str_value,
                     const gin::ArrayBufferView& buffer) {
  assert(static_cast<int>(buffer.num_bytes()) >= 0);
  v8::Handle<v8::String> str = str_value->ToString();
  int num_bytes = str->WriteUtf8(
      reinterpret_cast<char*>(buffer.bytes()),
      static_cast<int>(buffer.num_bytes()),
      NULL,
      v8::String::NO_NULL_TERMINATION | v8::String::REPLACE_INVALID_UTF8);
  return num_bytes;
}

int Utf8Length(v8::Handle<v8::Value> str_value) {
  return str_value->ToString()->Utf8Length();
}

gin::WrapperInfo g_wrapper_info = { gin::kEmbedderNativeGin };

}  // namespace

const char Unicode::kModuleName[] = "mojo/bindings/js/unicode";

v8::Local<v8::Value> Unicode::GetModule(v8::Isolate* isolate) {
  gin::PerIsolateData* data = gin::PerIsolateData::From(isolate);
  v8::Local<v8::ObjectTemplate> templ = data->GetObjectTemplate(
      &g_wrapper_info);

  if (templ.IsEmpty()) {
    templ = gin::ObjectTemplateBuilder(isolate)
                .SetMethod("decodeUtf8String", DecodeUtf8String)
                .SetMethod("encodeUtf8String", EncodeUtf8String)
                .SetMethod("utf8Length", Utf8Length)
                .Build();

    data->SetObjectTemplate(&g_wrapper_info, templ);
  }

  return templ->NewInstance();
}

}  // namespace js
}  // namespace mojo

File: mojo/public/js/bindings/unicode.js
// Copyright 2014 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

// Module "mojo/public/js/bindings/unicode"
//
// Note: This file is for documentation purposes only. The code here is not
// actually executed. The real module is implemented natively in Mojo.

while (1);

/**
 * Decodes the UTF8 string from the given buffer.
 * @param {ArrayBufferView} buffer The buffer containing UTF8 string data.
 * @return {string} The corresponding JavaScript string.
 */
function decodeUtf8String(buffer) { [native code] }

/**
 * Encodes the given JavaScript string into UTF8.
 * @param {string} str The string to encode.
 * @param {ArrayBufferView} outputBuffer The buffer to contain the result.
 * Should be pre-allocated to hold enough space. Use |utf8Length| to determine
 * how much space is required.
 * @return {number} The number of bytes written to |outputBuffer|.
 */
function encodeUtf8String(str, outputBuffer) { [native code] }

/**
 * Returns the number of bytes that a UTF8 encoding of the JavaScript string
 * |str| would occupy.
 */
function utf8Length(str) { [native code] }
